### PR TITLE
persistent groupings

### PR DESCRIPTION
**Pull request by FalkF**
## Related Issue

#206

## Description

~~incomplete~~ implementation with demo.
~~Bug: Table with different persistence-id still loads the saved groupings.~~

## Impacted Areas in Application

GroupBar
MaterialTable

## Additional Notes

I wanted to share my initial approach so you can comment on it :)
The deep copy of the columns array feels hacky, but that is probably just a JS thing...

### CODE REVIEW

**Comment by matthewoestreich** [comment]
Hi! Can we get the conflicts resolved on this?

**Comment by matthewoestreich** [comment]
I finally had some time to review this, somewhat "in-depth", but I really, really like this idea. It's an incredible idea!

The first thing that popped into my head, which may be dumb, is using cookies - since that will be "under the hood" - do we need to ask for permission to save cookies? Idk. Seems like a minor concern, just wanted to mention it.

I think this idea could be expanded upon (we have seen multiple issues where people using remote data want to "cache" how someone has selected rows, etc... on a previous page) so I think there's something more in your idea.

Do you have any opinion on making this "caching via cookies" more of a stand-out feature? What I mean by that is, imagine if we had some abstract way that would allow for the consumer to easily cache things that people seem to want cached. Or are we good with just overriding for now?

Great idea and implementation! Thanks for this.

**Comment by FalkF** [comment]
Glad you like it!
I actually did choose local storage over cookies, as it is more suited for caching user data (far easier to use). The legal implications of using local storage or cookies are the same. GDPR applies to both and requires user consent. I rarely saw it in the wild, but here is a german example of a privacy policy mentioning local storage: https://www.zdf.de/dsgvo/datenschutz-156.html

I think making sure that whoever reads the docs about cache related features gets the hint that local storage is used and GDPR or local laws may apply, is enough. Consent should be handled by other librarys, maybe we can link to some popular ones in the docs.

Yeah I think row selection could be saved in a similar way. Give the table an ID, save selected row IDs and check all matching rows once data is there. Oh and clean on page reload / session end would be good ofc.
But I think building something like a central store to handle storing and loading data would be to much.
Using local storage is a one liner. We could collect savedGroupings and savedSelections in an array within the local storage to keep it a little cleaner.
```
{
  material-table-core-groupings: [{ id, data }, { id, data }],
  material-table-core-selections: [{ id, data }, { id, data }],
  other-local-storage-enty: 'foo',
}
```

Sending the local storage data to a server and loading it on different clients (i.e. user logs in on different pc) could also be easily done outside of this package.

**Comment by matthewoestreich** [comment]
@FalkF thank you! My fault, I should have realized you were using local storage.  Thanks for your feedback, I really appreciate the help.

@Domino987 what do you think? Are you good with merging this? I think adding this component improves things.

File: src/components/MTableGroupbar/index.js [hunk 1]
```diff
@@ -32,6 +32,26 @@ function MTableGroupbar(props) {
     borderStyle: 'dashed'
   });
 
+  useEffect(() => {
+    if (props.persistentGroupingsId) {
+      const persistentGroupings = [];
+
+      props.groupColumns.forEach((column) => {
```
Comment (Domino987) [inline src/components/MTableGroupbar/index.js]: Can we exhacnge that for a map persistentGroupings = props.groupColumns.map(....)? To make the intentions clearer

File: src/components/MTableGroupbar/index.js [hunk 2]
```diff
@@ -32,6 +32,26 @@ function MTableGroupbar(props) {
...
+        persistentGroupings.push({
+          field: column.field,
+          groupOrder: column.tableData.groupOrder,
+          groupSort: column.tableData.groupSort,
+          columnOrder: column.tableData.columnOrder
+        });
+      });
+
+      localStorage.setItem(
+        props.persistentGroupingsId,
```
Comment (Domino987) [inline src/components/MTableGroupbar/index.js]: Do we want to pretag that with material table or something to reduce friction with other possible items?

File: src/material-table.js [hunk 3]
```diff
@@ -82,7 +82,35 @@ export default class MaterialTable extends React.Component {
           : '';
     }
 
-    this.dataManager.setColumns(props.columns, prevColumns);
+    const columnsCopy = JSON.parse(JSON.stringify(props.columns));
```
Comment (Domino987) [inline src/material-table.js:85]: Can we do the copy inside the if to reduce work? And wrap it with an else if not?

**Review by Domino987 — CHANGES_REQUESTED** [review]
I added a few comments. Can we change ti and merge it?

File: src/material-table.js (hunk 3 above)
Comment (FalkF) [reply src/material-table.js:85]: This solves an unreported? bug with shared column defintions and groupings :)

https://user-images.githubusercontent.com/19377375/125429648-f4851150-698e-4066-ae95-47af38d13d55.mov


```tsx
const columns = [{
    title: 'a',
    field: 'a',
},{
    title: 'b',
    field: 'b',
}];
const data = [{a: 'a', b:'b'}]
return (
<MaterialTable
          title="Basic"
          columns={columns}
          data={data}
          options={{
              grouping: true,
          }}
      />
      <MaterialTable
          title="Basic"
          columns={columns}
          data={data}
          options={{
              grouping: true,
          }}
      />)
```

File: src/components/MTableGroupbar/index.js (hunk 2 above)
Comment (FalkF) [reply src/components/MTableGroupbar/index.js]: I aggregate the groupings under material-table-groupings now

File: src/components/MTableGroupbar/index.js (hunk 1 above)
Comment (FalkF) [reply src/components/MTableGroupbar/index.js]: sure, done 👍

**Comment by FalkF** [comment]
@Domino987 Bump :)

**Comment by matthewoestreich** [comment]
@Domino987 thoughts on this?

**Comment by Domino987** [comment]
Technicality looks good, but there might be some side effects we are not aware yet. SO yes, lets merge it and add it to the website as well.

**Review by matthewoestreich — APPROVED** [review]
Thank you!

**Comment by FalkF** [comment]
I dont want to stress anyone, but is anything else needed from my side or can I help by adding docs to the site? :)

**Review by Domino987 — APPROVED** [review]
I think this looks good. once this is merged, we need to add it to the website though

**Pull request merged**